src/cbmc/cbmc_parse_options.cpp

```diff
@@ -526,6 +526,12 @@ int cbmc_parse_optionst::doit()
   if(set_properties())
     return CPROVER_EXIT_SET_PROPERTIES_FAILED;
 
+  if(cmdline.isset("validate-goto-model"))
```

Comment:
Would we like to have this in the other driver programs too?

Reply:
I would have thought goto-instrument was the place for it.

Reply:
Yes, I think we should have the option in all of the tools to validate the goto model.

@martin-cs I think also in e.g. `cbmc` it would make sense to validate the goto program as a safety net sort of thing. Once right after `goto_convert` and perhaps once more before `goto_symex` to check that all the transformation passes like `remove_function_pointers`, etc. did not produce an ill-formed goto model.

Reply:
@danpoe : I'm very much in favour of validating the program between all transformation passes; that sounds great.  To the point I might suggest making that the last step of each transformation.  I think this will need to be able to check for various "normal forms" as well, i.e. returns and function pointers removed, etc. ( @tautschnig may have opinions on this subject ).  As I first envisaged this, that should be enough to prove that all of the DATA_INVARIANTs hold.

i also like the "with exceptions when we load it and with invariants after" approach.  My only question is about the specific, user accessible command-line flag to just check the validity of the goto-program and nothing else.  That seems like something we definitely want but that it should probably be in goto-instrument as a "checking and modifying the goto-program" bit of functionality, rather than in every tool.

Reply:
Ah yes, that makes sense. @sonodtt is currently working on some additional checks of the goto program that will allow to pass flags that indicate whether function pointers and returns are expected to have been removed.

Reply:
Cool.

@sonodtt : having looked at this in the past ( https://github.com/diffblue/cbmc/issues/751 is the first bit I have written down ), may I suggest that you make the flags independent for each feature.  For example, one of the last few differences between the ```static_analyst``` and ```ait``` frameworks is that the later requires the abscence of function pointers (naturally or removed) and return removal.  function pointer removal is an (at the moment undocumented) precondition of return removal and so on.  It would be *amazing* if we could actually map out and enforce these requirements on "normal forms".